docs: updated README for App with information about the route generation process (#20586)
cypress-io/cypress: packages/app/README.md

3. Open chrome (or another browser)
4. It will show the new Vite powered app 

## How the App works

Cypress has two modes: `run` and `open`. We want run mode to be as light and fast as possible, since this is the mode used to run on CI machines, etc. Open mode is the interactive experience, with the command log, snapshots, selector playground, etc.

- **`open`** mode is driven using GraphQL and urql. It shows the full Cypress app, include the top nav, side nav, spec list, etc. You can change between testing types, check your latest runs on the Cypress dashboard, updating settings, etc.
- **`run`** mode is does not rely on GraphQL. This is so we can be as performant as possible. It only renders the "runner" part of the UI, which is comprised of the command log, Spec Runner header, and AUT iframe.

The two modes are composed using the same logic, but have slightly different components. You can see where the differences are in `Runner.vue`(src/pages/Specs/Runner.vue). Notice that `<SpecRunnerOpenMode>` receives a `gql` prop, since it uses GraphQL, and `<SpecRunnerRunMode>` does not.

## Router

> [`vite-plugin-pages`](https://github.com/hannoeru/vite-plugin-pages) is used to generate routes based on the page-level Vue components contained in `src/pages`. Route configuration that might typically appear in `router.ts` can be set in a `<route>` block in these page components, for example:

Reviewer: It might be worth noting that these options (name, meta) directly map to Vue Router's options and link to that? It took me a while to make this connection.

Author: > One comment, can we link to Vue Router somewhere? It might not be clear that is the underlying routing mechanism (it is, right?)

We link at the bottom, but I agree it would be more clear to mention up top especially if folks aren't familiar with Vue and Vue Router. Will add an intro to connect the dots between the Vite plugin, the `<route>` block in SFC and Vue Router's options.

Author: Changes made!

3. Open chrome (or another browser)
4. It will show the new Vite powered app 

> 

Reviewer 2: Nit: the doc says this is the front-end for the Cypress App, however the Launchpad and reporter packages also contain front-end for Cypress. Can we make this a bit more specific to what this package is accomplishing vs those packages?

Author: We should do that, but let's approach it another PR. I'd love to stop referring to this package as "app" when there are 4 packages (the ones you listed and frontend-shared) that comprise the App from a user perspective.

Reviewer 2: It'd be great if we could actually co-locate these someday
```
packages/app
      src/
         reporter/
         launchpad/
         shared/
         ...app specific files

```

Reviewer: The dream

## How the App works

> Cypress has two modes: `run` and `open`. We want run mode to be as light and fast as possible, since this is the mode used to run on CI machines, etc. Open mode is the interactive experience, with the command log, snapshots, selector playground, etc.

Author: Cleaning this up a little.

- **`open`** mode is driven using GraphQL and urql. It shows the full Cypress app, include the top nav, side nav, spec list, etc. You can change between testing types, check your latest runs on the Cypress dashboard, updating settings, etc.
- **`run`** mode is does not rely on GraphQL. This is so we can be as performant as possible. It only renders the "runner" part of the UI, which is comprised of the command log, Spec Runner header, and AUT iframe.

The two modes are composed using the same logic, but have slightly different components. You can see where the differences are in `Runner.vue`(src/pages/Specs/Runner.vue). Notice that `<SpecRunnerOpenMode>` receives a `gql` prop, since it uses GraphQL, and `<SpecRunnerRunMode>` does not.

## Router

> The App's routing doesn't need to be touched often, because it is almost all auto-generated based on the file structure. There's so little code that it helps to describe the approach here so that when we do want to modify a route or do some other route-specific behavior, we can get our bearings.

Reviewer 2: Can you link to the grunt script that handles this generation?

Author: There's no script, it's `vite-plugin-pages` that does the work, using [virtual modules](https://vitejs.dev/guide/api-plugin.html#virtual-modules-convention), so in `router.ts` we can do these imports:

```
import generatedRoutes from 'virtual:generated-pages'
import { setupLayouts } from 'virtual:generated-layouts'
```

> [`vite-plugin-pages`](https://github.com/hannoeru/vite-plugin-pages) is used to generate routes based on the page-level Vue components contained in `src/pages`. These generated routes are pulled into a standard [Vue Router](https://router.vuejs.org/) setup using `createRouter()` in `router.ts`.

Reviewer 2: Can we link to `router.ts`?

Route configuration that might typically appear in `router.ts` can be set in a `<route>` block in these page components, for example `name` and `meta` properties (documented in Vue Router docs):

```ts
<route>
  {
    name: 'SpecsRunner',
    meta: {
      header: false,
      navBarExpandedAllowed: false
    }
  }
</route>
```

> The advantage here is that the route definition is co-located with the component. The route block will be parsed as JSON5 by default, which means that certain Vue Router properties aren't valid in the `<route>` block, such as if a function is needed to be the value. 

Reviewer 2: > such as if a function is needed to be the value.

is this saying func props arn't allowed? Is this a limitation with how we are auto-generating these routes or a JSON5 limitation or a limitation with Vue?

So far, we haven't needed to use any such properties, but if we do, [`extendRoute`](https://github.com/hannoeru/vite-plugin-pages#extendroute) can be used where the `Pages` plugin is added in `vite.config.ts` to extend the generated route objects, using regular TypeScript. 

### Some gotchas if you are new to Vue Router

#### `name` vs `path`

With the implicit setup of the router, routes are named based on the file name. This is not always ideal as the route's name is used in the UI as the title. The `name` property can be overridden in the `<route>` block of a single file component.

> In addition to sometimes showing a route's name in the UI, we also use the `name` value to push a new route object to the router (either in code directly, or using `RouterLink`s), if that object includes `params`. Otherwise we usually use `path` when pushing to the router.

Reviewer 2: > we also use the `name` value to push a new route object to the router (either in code directly, or using `RouterLink`s),

Can you link to an example where we've had to do this?

Author: I'm a bit reluctant to link to much specific code as it's so likely to get moved around and changed. I can add some inline examples though.

#### `params` vs `query`

> We use `params` when there is temporary state that should be present when a user gets to a route from somewhere else in the App, but shouldn't be present after page refresh or when using the `back` button. Useful for things like the notice if you view a spec right after creating it, for example.

Reviewer 2: ```suggestion
We use `params` when there is temporary state that should be present when a user gets to a route from somewhere else in the App, but shouldn't be present after page refresh or when using the `back` button. For example, this is useful for showing a user a notification to view a spec right after creating it.
```

Author: This suggestion doesn't quite have the same meaning, the notice is _after_ you view the spec, but I'll take it as feedback to make the sentence clearer.

> For values that should survive a page refresh, which is most cases where we'd put something in the route, we use `query` in the route object to put the values into the URL's query parameters.

Reviewer 2: ```suggestion
For values that should survive a page refresh, which is most cases, we use `query` in the route object to put the values into the URL's query parameters.
```
